driver-core/src/main/com/mongodb/internal/connection/DefaultServerMonitor.java

```diff
@@ -272,27 +344,88 @@ static boolean shouldLogStageChange(final ServerDescription previous, final Serv
             return true;
         }
         if (previous.getElectionId() != null
-                    ? !previous.getElectionId().equals(current.getElectionId()) : current.getElectionId() != null) {
+                ? !previous.getElectionId().equals(current.getElectionId()) : current.getElectionId() != null) {
             return true;
         }
         if (previous.getSetVersion() != null
-                    ? !previous.getSetVersion().equals(current.getSetVersion()) : current.getSetVersion() != null) {
+                ? !previous.getSetVersion().equals(current.getSetVersion()) : current.getSetVersion() != null) {
             return true;
         }
 
         // Compare class equality and message as exceptions rarely override equals
         Class<?> thisExceptionClass = previous.getException() != null ? previous.getException().getClass() : null;
         Class<?> thatExceptionClass = current.getException() != null ? current.getException().getClass() : null;
-        if (thisExceptionClass != null ? !thisExceptionClass.equals(thatExceptionClass) : thatExceptionClass != null) {
+        if (!Objects.equals(thisExceptionClass, thatExceptionClass)) {
             return true;
         }
 
         String thisExceptionMessage = previous.getException() != null ? previous.getException().getMessage() : null;
         String thatExceptionMessage = current.getException() != null ? current.getException().getMessage() : null;
-        if (thisExceptionMessage != null ? !thisExceptionMessage.equals(thatExceptionMessage) : thatExceptionMessage != null) {
+        if (!Objects.equals(thisExceptionMessage, thatExceptionMessage)) {
             return true;
         }
 
         return false;
     }
+
+
+    private class RoundTripTimeRunnable implements Runnable {
+        private volatile InternalConnection connection = null;
+
+        void close() {
+            InternalConnection connection = this.connection;
+            if (connection != null) {
+                connection.close();
+            }
+        }
+
+        @Override
+        public void run() {
+            try {
+                while (!isClosed) {
+                    try {
+                        if (connection == null) {
+                            initialize();
+                        } else {
+                            try {
+                                pingServer(connection);
+                            } catch (MongoSocketException | MongoCommandException e) {
+                                connection.close();
+                                connection = null;
```

Comment:
Yes, this is a bug.  But instead of throwing e I instead removed this entire catch block and close the connection in the outer catch block.  This way if `connection.open` throws an exception, the connection will be set to null as well and re-created the next time around the loop.